rp2040-hal/src/adc.rs

```diff
@@ -153,3 +154,210 @@ where
         Ok(self.device.result.read().result().bits().into())
     }
 }
+
+impl Adc {
+    pub fn free_running<'a>(&'a mut self) -> FreeRunning<'a> {
+        FreeRunning { adc: self }
+    }
+}
+
+pub struct FreeRunning<'a> {
+    adc: &'a mut Adc,
+}
+
+impl<'a> FreeRunning<'a> {
+    /// Manually set clock divider integral and fractional parts
+    pub fn clock_divider(self, int: u16, frac: u8) -> Self {
+        self.adc.device.div.modify(|_, w| unsafe { w.int().bits(int).frac().bits(frac) });
+        self
+    }
+
+    /// Select initial ADC input to sample from
+    ///
+    /// If round-robin mode is used, this will only affect the first sample.
+    pub fn initial_input<PIN: Channel<Adc, ID = u8>>(self, _pin: &mut PIN) -> Self {
+        self.adc.device.cs.modify(|_, w| unsafe { w.ainsel().bits(PIN::channel()) });
+        self
+    }
+
+    /// Set channels to use for round-robin mode
+    ///
+    /// Takes a tuple of channels, like `(&mut adc_pin, &mut temp_sense)`.
+    ///
+    /// The order in which the channels are specified has no effect.
+    /// Channels are always sampled in increasing order (Channel 0, Channel 1, ...).
+    pub fn round_robin<T: Into<RoundRobin>>(self, selected_channels: T) -> Self {
+        let RoundRobin(bits) = selected_channels.into();
+        self.adc.device.cs.modify(|_, w| unsafe { w.rrobin().bits(bits) });
+        self
+    }
+
+    /// Enable ADC fifo and start free-running conversion
+    ///
+    /// Use the returned `Fifo` instance to access the captured data.
+    ///
+    /// To stop capturing, call `fifo.stop()`.
+    pub fn start_fifo(self) -> Fifo<'a, false> {
+        self.adc.device.fcs.modify(|_, w| w.en().set_bit());
+        self.adc.device.cs.modify(|_, w| w.start_many().set_bit());
+        Fifo { adc: self.adc }
+    }
+
+    /// Enable ADC fifo with DMA and start free-running conversion
+    ///
+    /// In addition to being usable the same way as a `Fifo` returned from `start_fifo`,
+    /// the Fifo returned by this function can also be used as a source for DMA transfers.
+    pub fn start_fifo_with_dma(self, thresh: u8) -> Fifo<'a, true> {
+        self.adc.device.fcs.modify(|_, w| unsafe {
+            w.en().set_bit()
+                .dreq_en().set_bit()
+                .thresh().bits(thresh)
+        });
+        self.adc.device.cs.modify(|_, w| w.start_many().set_bit());
```

Comment:
This doesn't adhere to the documented sequence for starting DMA on the ADC:

> Once the ADC is suitably configured, the DMA channel should be started first, and the ADC conversion should be started second, via CS.START_MANY. Once the DMA completes, the ADC can be halted, or a new DMA transfer promptly started. After clearing CS.START_MANY to halt the ADC, software should also poll CS.READY to make sure the last conversion has finished, and then drain any stray samples from the FIFO.

This is something we probably can't do properly with the current DMA API.
Not sure what to do about it. If you tested your code and it works reliably, perhaps we can just add a doc comment that this isn't properly following the spec for now?

Reply:
tbh, I know very little about the DMA part, and I have not tested it thoroughly yet. As mentioned above, maybe it's better to drop it for now.